utils/url/url.go

```diff
@@ -0,0 +1,136 @@
+package url
+
+import (
+	"strings"
+
+	"github.com/ARM-software/golang-utils/utils/commonerrors"
+	"github.com/ARM-software/golang-utils/utils/reflection"
+)
+
+const defaultPathSeparator = "/"
+
+// IsParamSegment checks whether the segment string is a path parameter
+func IsParamSegment(segment string) bool {
+	return len(segment) >= 2 && segment[0] == '{' && segment[len(segment)-1] == '}'
+}
+
+// HasMatchingPathSegments checks whether two path strings match based on their segments.
+func HasMatchingPathSegments(pathA, pathB string) (bool, error) {
+	return hasMatchingPathSegments(pathA, pathB, func(segmentA, segmentB string) bool {
+		return segmentA == segmentB
+	})
+}
+
+// HasMatchingPathSegmentsWithParams is similar to HasMatchingPathSegments but also considers segments as matching if at least one of them contains a path parameter.
+//
+//	HasMatchingPathSegmentsWithParams("/some/{param}/path", "/some/{param}/path") // true
+//	HasMatchingPathSegmentsWithParams("/some/abc/path", "/some/{param}/path") // true
+//	HasMatchingPathSegmentsWithParams("/some/abc/path", "/some/def/path") // false
+func HasMatchingPathSegmentsWithParams(pathA, pathB string) (bool, error) {
+	return hasMatchingPathSegments(pathA, pathB, func(pathASeg, pathBSeg string) bool {
+		switch {
+		case IsParamSegment(pathASeg):
+			return pathBSeg != ""
+		case IsParamSegment(pathBSeg):
+			return pathASeg != ""
+		default:
+			return pathASeg == pathBSeg
```

Comment:
I think we need to be a bit more thorough and use the url.PathUnescape(path) on each segment in case they are encoded